src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -344,133 +344,303 @@ public class SqlMutationEngine : IMutationEngine
+                bool isDatabasePolicyDefinedForReadAction = false;
 
-                try
+                if (isReadPermissionConfiguredForRole)
                 {
-                    // Creating an implicit transaction
-                    using (TransactionScope transactionScope = ConstructTransactionScopeBasedOnDbType())
-                    {
-                        upsertOperationResult = await PerformUpsertOperation(
-                                                            parameters,
-                                                            context);
-                        transactionScope.Complete();
-                    }
+                    isDatabasePolicyDefinedForReadAction = !string.IsNullOrWhiteSpace(_authorizationResolver.GetDBPolicyForRequest(context.EntityName, roleName, EntityActionOperation.Read));
                 }
 
-                // All the exceptions that can be thrown by .Complete() and .Dispose() methods of transactionScope
-                // derive from TransactionException. Hence, TransactionException acts as a catch-all.
-                // When an exception related to Transactions is encountered, the mutation is deemed unsuccessful and
-                // a DataApiBuilderException is thrown
-                catch (TransactionException)
+                if (context.OperationType is EntityActionOperation.Upsert || context.OperationType is EntityActionOperation.UpsertIncremental)
                 {
-                    throw _dabExceptionWithTransactionErrorMessage;
-                }
+                    DbResultSet? upsertOperationResult;
+                    DbResultSetRow? upsertOperationResultSetRow = null;
+                    JsonDocument? selectOperationResponse = null;
+                    bool isUpdateResultSet = false;
 
-                DbResultSetRow? dbResultSetRow = upsertOperationResult is not null ?
-                    (upsertOperationResult.Rows.FirstOrDefault() ?? new()) : null;
+                    try
+                    {
+                        // Creating an implicit transaction
+                        using (TransactionScope transactionScope = ConstructTransactionScopeBasedOnDbType())
+                        {
+                            upsertOperationResult = await PerformUpsertOperation(
+                                                                parameters,
+                                                                context);
 
-                if (upsertOperationResult is not null &&
-                    dbResultSetRow is not null && dbResultSetRow.Columns.Count > 0)
-                {
-                    Dictionary<string, object?> resultRow = dbResultSetRow.Columns;
+                            upsertOperationResultSetRow = upsertOperationResult is not null ? (upsertOperationResult.Rows.FirstOrDefault() ?? new())
```

Comment:
since `upsertOperationResult` is never null, we probably don't need conditional operator expression. Perhaps an if/else , where if upsertOperationResult is null, an exception is thrown (this would actually be an exceptional circumstance because of the claim it will never be null.) Since you're already technically checking for null here, you can add the else case/exception so you can avoid using the null forgiving operator `!` a few lines down.

Reply:
I've added logic to throw exception when the resultset is null. Also, re-wrote some conditionals here and in insert operation flow which would allow skipping the usage of null forgiving operator.